FirebaseAuth/Sources/Utilities/FIRAuthRecaptchaVerifier.m

```diff
@@ -0,0 +1,239 @@
+  if (!forceRefresh) {
+    if ([FIRAuth auth].tenantID == nil && _agentConfig != nil) {
+      completion(nil);
+      return;
+    }
+    if ([FIRAuth auth].tenantID != nil && _tenantConfigs[[FIRAuth auth].tenantID] != nil) {
+      completion(nil);
+      return;
+    }
+  }
+  FIRGetRecaptchaConfigRequest *request = [[FIRGetRecaptchaConfigRequest alloc]
+      initWithRequestConfiguration:[FIRAuth auth].requestConfiguration];
+  [FIRAuthBackend
+      getRecaptchaConfig:request
+                callback:^(FIRGetRecaptchaConfigResponse *_Nullable response,
+                           NSError *_Nullable error) {
+                  if (error) {
+                    completion(error);
+                  }
+                  FIRAuthRecaptchaConfig *config = [[FIRAuthRecaptchaConfig alloc] init];
+                  config.siteKey = [response.recaptchaKey componentsSeparatedByString:@"/"][3];
+
+                  // TODO(chuanr@): retrieve provider enablement status when backend is ready
+                  if (!forceRefresh) {
+                    config.enablementStatus = @{@"password" : @NO};
+                  } else {
+                    config.enablementStatus = @{@"password" : @YES};
+                  }
+
+                  if ([FIRAuth auth].tenantID == nil) {
+                    self->_agentConfig = config;
+                    completion(nil);
+                    return;
+                  } else {
+                    if (!self->_tenantConfigs) {
+                      self->_tenantConfigs = [[NSMutableDictionary alloc] init];
+                    }
+                    self->_tenantConfigs[[FIRAuth auth].tenantID] = config;
+                    completion(nil);
+                    return;
+                  }
+                }];
+}
+
+- (void)retrieveRecaptchaTokenWithAction:(FIRAuthRecaptchaAction)action
+                              completion:(nullable FIRAuthRecaptchaTokenCallback)completion {
+#if TARGET_OS_IOS && !TARGET_OS_MACCATALYST
+  Class RecaptchaActionClass = NSClassFromString(@"RecaptchaAction");
+  if (RecaptchaActionClass) {
+    SEL customActionSelector = NSSelectorFromString(@"initWithCustomAction:");
+
+    if ([RecaptchaActionClass instancesRespondToSelector:customActionSelector]) {
+      // Initialize with a custom action
+      id (*funcWithCustomAction)(id, SEL, NSString *) =
+          (id(*)(id, SEL,
+                 NSString *))[RecaptchaActionClass instanceMethodForSelector:customActionSelector];
+
+      id<RCAActionProtocol> customAction = funcWithCustomAction(
+          [[RecaptchaActionClass alloc] init], customActionSelector, actionToStringMap[@(action)]);
+
+      if (customAction) {
+        NSLog(@"Created custom action");
+        [self.recaptchaClient execute:customAction
+                           completion:^(NSString *_Nullable token, NSError *_Nullable error) {
+                             completion(token, error);
+                             return;
+                           }];
```

Comment:
Conditionally link to recaptcha sdk with objc reflection, to execute with a custom action.